arm-software/embedded/llvmlibc-support/crt0/exceptions_7a.h

```diff
@@ -20,36 +20,67 @@ namespace exceptions {
 
 using namespace sysreg;
 
+// The following function is only valid in architectures with the
+// Virtualization Extension. armv7ve is the most basic architecture with it.
+// However, because armv7ve is not a valid value for the target attribute,
+// Cortex-A7 is used instead. The code is simple enough not to use instructions
+// that would be valid for Cortex-A7 but not for armv7ve.
+[[gnu::target("arch=cortex-a7")]] static uintptr_t read_elr_hyp() {
+  return __arm_rsr("ELR_hyp");
+}
+
+static bool is_in_hyp_mode() {
+  const uint32_t cpsr = __arm_rsr("CPSR");
+  const uint32_t mode_bits = cpsr & 0x1F;
+  return mode_bits == 0x1A;
+}
+
+#define GET_LR_VALUE_INTO(var)                                                 \
+  do {                                                                         \
+    if (is_in_hyp_mode())                                                      \
+      var = read_elr_hyp();                                                    \
+    else                                                                       \
+      var = (uint32_t)__builtin_return_address(0);                             \
+  } while (0)
```

Comment:
Is there a reason you didn't define this macro as an expression rather than a statement, along the lines of this?
```c
#define GET_LR_VALUE() (is_in_hyp_mode() ? read_elr_hyp() : (uintptr_t)__builtin_return_address(0))
```

Then you wouldn't have had to change all those one-line call sites `print_hex(...)` into three lines with a variable declaration.

Reply:
Excellent suggestion. Much better now.